### PR TITLE
Drop log unsupported "spdy", make "http2" a server level option.

### PR DESCRIPTION
**Pull request by oxpa**
### Proposed changes

Removes long abandoned SPDY listening socket option.
Also moves http2 to a server level 

### Checklist

Before creating a PR, run through this checklist and mark each as complete.

- [x] I have read the [`CONTRIBUTING`](https://github.com/nginxinc/ansible-role-nginx-config/blob/main/CONTRIBUTING.md) document
- [x] I have added Molecule tests that prove my fix is effective or that my feature works
- [x] I have checked that any relevant Molecule tests pass after adding my changes
- [x] I have updated any relevant documentation ([`defaults/main/*.yml`](https://github.com/nginxinc/ansible-role-nginx-config/blob/main/defaults/main/), [`README.md`](https://github.com/nginxinc/ansible-role-nginx-config/blob/main/README.md) and [`CHANGELOG.md`](https://github.com/nginxinc/ansible-role-nginx-config/blob/main/CHANGELOG.md))

### CODE REVIEW

**Comment by oxpa** [comment]
commit message is wrong >_<

**Review by alessfg — CHANGES_REQUESTED** [review]
Thanks for the PR! I missed `spdy` no longer being supported.

Re implementing the new `http2` directives -- the way the templates are organized, each "module" has its own template. Your current placement is within the `core` template, whereas the `http2` directive should live inside its own module template. This helps in maintaining and tracking down any potential issues with the code further down the line. I think the best location for the module would be before this macro block here https://github.com/nginxinc/ansible-role-nginx-config/blob/main/templates/http/modules.j2#L23. You will then need to add the template to https://github.com/nginxinc/ansible-role-nginx-config/blob/main/templates/http/default.conf.j2. (There is no need to implement all the `http2` directives right now, but creating its own template section will help with implementing the remaining directives further down the line 😄)

**Comment by oxpa** [comment]
OK, will do.
It's also better to postpone merging this PR till R30, when these changes are in N+

File: defaults/main/template.yml [hunk 1]
```diff
@@ -273,6 +271,13 @@ nginx_config_http_template:
         underscores_in_headers: false  # Boolean -- Not available in the 'location' context
         variables_hash_bucket_size: 64  # Available only in the 'http' context
         variables_hash_max_size: 1024  # Available only in the 'http' context
+      http2: # all options are for 'http' or 'server' contexts only unless noted otherwise
+        enable: false
+        body_preread_size: 64k
+        chunk_size: 8k # can be used in a 'location' context
```
Comment (alessfg) [inline defaults/main/template.yml:277]:
```suggestion
        chunk_size: 8k  # Not available in the 'location' context
```

File: defaults/main/template.yml [hunk 2]
```diff
@@ -273,6 +271,13 @@ nginx_config_http_template:
...
+        max_concurrent_streams: 128
+        recv_buffer_size: 256k # 'http' level only
```
Comment (alessfg) [inline defaults/main/template.yml:279]:
```suggestion
        recv_buffer_size: 256k  # Available only in the 'http' context
```

File: defaults/main/template.yml [hunk 3]
```diff
@@ -273,6 +271,13 @@ nginx_config_http_template:
         underscores_in_headers: false  # Boolean -- Not available in the 'location' context
         variables_hash_bucket_size: 64  # Available only in the 'http' context
         variables_hash_max_size: 1024  # Available only in the 'http' context
+      http2: # all options are for 'http' or 'server' contexts only unless noted otherwise
```
Comment (alessfg) [inline defaults/main/template.yml:274]:
```suggestion
      http2:  # Configure HTTP2
```

File: molecule/default/converge.yml [hunk 4]
```diff
@@ -653,6 +661,8 @@
                       core:
                         index: frontend_index.html
                         root: /usr/share/nginx/html
+                      http2:
+                        chunk_size: 8k
```
Comment (alessfg) [inline molecule/default/converge.yml:665]: Lets test a different value to ensure the defaults are being correctly replaced
```suggestion
                        chunk_size: 10k
```

File: templates/http/modules.j2 [hunk 5]
```diff
@@ -330,3 +330,26 @@ sub_filter_types {{ sub_filter['types'] if sub_filter['types'] is string else su
...
+{% endif %}
+{% endif %}
+{% if scope == 'http' and http2['recv_buffer_size'] is defined %} {# can only be in 'http' context #}
+http2_recv_buffer_size {{ http2['recv_buffer_size'] }};
+{% endif %}
+{% if http2['chunk_size'] is defined %} {# can be wherever #}
+http2_chunk_size {{ http2['chunk_size'] }};
+{% endif %}
+{% endif %}
+{% endmacro%}
```
Comment (alessfg) [inline templates/http/modules.j2:355]: Due to Jinja2 being somewhat finicky in formatting, it's best to leave an empty line before the `{% endmacro %}` parameter
```suggestion

{% endmacro %}
```

File: templates/http/modules.j2 [hunk 6]
```diff
@@ -330,3 +330,26 @@ sub_filter_types {{ sub_filter['types'] if sub_filter['types'] is string else su
...
+http2_body_preread_size {{ http2['body_preread_size'] }};
+{% endif %}
+{% if http2['max_concurrent_streams'] is defined %}
+http2_max_concurrent_streams {{ http2['max_concurrent_streams'] }};
+{% endif %}
+{% endif %}
+{% if scope == 'http' and http2['recv_buffer_size'] is defined %} {# can only be in 'http' context #}
+http2_recv_buffer_size {{ http2['recv_buffer_size'] }};
+{% endif %}
+{% if http2['chunk_size'] is defined %} {# can be wherever #}
```
Comment (alessfg) [inline templates/http/modules.j2:351]: The general assumption is that directives are allowed everywhere, there is no need to specify
```suggestion
{% if http2['chunk_size'] is defined %}
```

File: templates/http/modules.j2 [hunk 7]
```diff
@@ -330,3 +330,26 @@ sub_filter_types {{ sub_filter['types'] if sub_filter['types'] is string else su
...
+http2 {{ http2['enabled'] | ternary('on', 'off') }};
+{% endif %}
+{% if http2['body_preread_size'] is defined %}
+http2_body_preread_size {{ http2['body_preread_size'] }};
+{% endif %}
+{% if http2['max_concurrent_streams'] is defined %}
+http2_max_concurrent_streams {{ http2['max_concurrent_streams'] }};
+{% endif %}
+{% endif %}
+{% if scope == 'http' and http2['recv_buffer_size'] is defined %} {# can only be in 'http' context #}
```
Comment (alessfg) [inline templates/http/modules.j2:348]:
```suggestion
{% if scope == 'http' and http2['recv_buffer_size'] is defined %}{# 'recv_buffer_size' directive is only available in the 'http' context #}
```

File: templates/http/modules.j2 [hunk 8]
```diff
@@ -330,3 +330,26 @@ sub_filter_types {{ sub_filter['types'] if sub_filter['types'] is string else su
 {% endif %}
 
 {% endmacro %}
+
+{# NGINX HTTP v2 -- ngx_http_v2_module #}
+{% macro http2(http2, scope='http') %}
+{% if http2 is defined %}
+{% if scope != 'location' %} {# 'http' and 'server' contexts #}
```
Comment (alessfg) [inline templates/http/modules.j2:337]: Leaving an empty space between the Jinja2 templated line and the Jinja2 comment will lead to some "undesired" white space in the NGINX config
```suggestion
{% if scope != 'location' %}{# The following scoped directives are not available in the 'location' context #}
```

**Review by alessfg — CHANGES_REQUESTED** [review]
Looks great! I added a few suggestions re the wording in comments (to keep all comments as standard as possible) and fix some Jinja2 quirks that might lead to unnecessary whitespaces.

One more thing, can you update the Changelog? Thanks!

File: defaults/main/template.yml (hunk 1 above)
Comment (oxpa) [reply defaults/main/template.yml:277]: This change inverts the meaning of the comment. Chunk size is the only option that can be used in a location.

Comment (oxpa) [reply defaults/main/template.yml:277]: It's easier to drop the comment, I guess. Along with the main one.

**Review by alessfg — APPROVED** [review]

**Pull request merged**